### PR TITLE
fix hard coded device_id

### PR DESCRIPTION
**Pull request by HectorSVC**
**Description**: Describe your changes.
There's are some place hard code the device_id to 0 while using AsyncBuffer which makes it run into issue user want to use device other than device_id 0.

https://github.com/microsoft/onnxruntime/issues/1592

### CODE REVIEW

File: onnxruntime/core/framework/graph_partitioner.cc [hunk 1]
```diff
@@ -12,7 +12,7 @@
 #include "core/framework/func_kernel.h"
 
 // uncomment this line to count non-CUDA ops in ONNX domain
-//#define COUNT_NON_CUDA_OPS
+#define COUNT_NON_CUDA_OPS
```
Comment (KeDengMS) [inline onnxruntime/core/framework/graph_partitioner.cc]: unintended change?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 2]
```diff
@@ -70,7 +70,7 @@ CUDAExecutionProvider::CUDAExecutionProvider(const CUDAExecutionProviderInfo& in
   InsertAllocator(CreateAllocator(default_allocator_info, device_id_));
 
   DeviceAllocatorRegistrationInfo pinned_allocator_info(
-      {OrtMemTypeCPUOutput, [](int) { return std::make_unique<CUDAPinnedAllocator>(0, CUDA_PINNED); }, std::numeric_limits<size_t>::max()});
+      {OrtMemTypeCPUOutput, [](int id) { return std::make_unique<CUDAPinnedAllocator>(id, CUDA_PINNED); }, std::numeric_limits<size_t>::max()});
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: Pinned allocator is in CPU, so no device id needed?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 3]
```diff
@@ -80,7 +80,7 @@ CUDAExecutionProvider::CUDAExecutionProvider(const CUDAExecutionProviderInfo& in
 }
 
 CUDAExecutionProvider::~CUDAExecutionProvider() {
-  auto cpu_alloc = GetAllocator(0, OrtMemTypeCPU);
+  auto cpu_alloc = GetAllocator(device_id_, OrtMemTypeCPU);
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: >auto cpu_alloc = GetAllocator(device_id_, OrtMemTypeCPU); [](start = 2, length = 57)

why cpu_alloc needs GPU device_id?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 2 above)
Comment (HectorSVC) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: The major reason is CudaAsyncBuffer need a device_id for CUDA allocator. There's more place need to be fixed. Will update later.

---
In reply to: [321858267](https://github.com/microsoft/onnxruntime/pull/1771#discussion_r321858267) [](ancestors = 321858267)

**Pull request closed without merging**